### ISSUE TITLE
PKI certificate issue fails as CA approaches expiration

### ISSUE THREAD

**Issue by jsmilani** [hashicorp/vault#13193]
**Describe the bug**
We use a pre-generated corporate CA cert in our Vault PKI backend root certificate. This root certificate is going to expire next year and we are discovering that some of our tools that need long lasting certs are failing to generate certificates with the error `cannot satisfy request, as TTL would result in notAfter 2022-... that is beyond the expiration of the CA certificate at 2022-...`. When we make the request to issue the certificate we are not actually requesting a specific end date so it defaults to the one calculated by Vault based on the TTL values for the given role. This seems to ignore the fact that the CA itself expires at a specific date. The problem is the endpoint calculates the expiration of the current cert from the current time + TTL but this is a moving target and is not easy for clients to calculate and explicitly ask for something before the CA expiration, but it would be trivial for Vault to calculate and cap the certificate based of the CA expiration.

The implication of this bug is that things that services that are working perfectly fine for months and years start to fail to renew certificates if they are unlucky enough to have default TTL longer than the time left on their certificates. Failures start occurring at (CA Cert expiration - TTL) instead of at (CA Cert expiration).

**To Reproduce**
1. generate a pki mount with max TTL of 60 days:
    ```
    vault secrets enable pki
    vault secrets tune -max-lease-ttl=1440h pki
    ```
2. generate a root cert that expires in 29 days
    ```
    vault write pki/root/generate/internal common_name=test ttl=696h
    ```
3. create a role with ttl 30d
    ```
    vault write pki/roles/example-dot-com allowed_domains=my-website.com allow_subdomains=true ttl=720h max_ttl=720h
    ```
4. request a cert
    ```
    vault write pki/issue/example-dot-com common_name=www.my-website.com
    ```
5. See error
    ```
    Error writing data to pki/issue/example-dot-com: Error making API request.

    URL: PUT http://127.0.0.1:8200/v1/pki/issue/example-dot-com
    Code: 400. Errors:

    * cannot satisfy request, as TTL would result in notAfter 2021-12-17T20:53:34.1477146Z that is beyond the expiration of the CA certificate at 2021-12-16T20:46:53Z
    ```

**Expected behavior**
I did not specify an explicit date so the endpoint automatically calculated one from the TTL which turned out to be invalid. It would be ideal if the logic takes into account the CA end date and automatically truncate the TTL to match. Even if I specify a TTL the expected behavior should still truncate it.

**Environment:**
* Vault Server Version (retrieve with `vault status`): Vault v1.8.4
* Vault CLI Version (retrieve with `vault version`): Vault v1.8.4
* Server Operating System/Architecture: Linux/Docker

Vault server configuration file(s):
NA, I started this within docker so this is easily repeatable inside a container `docker run --rm -d --cap-add=IPC_LOCK vault`

**Additional context**

**Comment by hsimon-hashicorp** [comment]
@jsmilani This is an interesting one! Thank you very much for the detailed report and repro steps. I'll check with our engineers on this and see if this is expected, or if we could possibly handle this a bit better. (Or possibly document this a bit better.) Thanks again!

**Comment by schrodingersket** [comment]
I'm encountering exactly the same use-case and corresponding workflow issues; it looks like 
this has been discussed a little previously here, but was never re-visited: https://github.com/hashicorp/vault/issues/5834

**Comment by ashwinkupatkar** [comment]
Yes.... this is an issue to me as well as I have to calculate the ttl for the derivative/leaf certs ... it's quite a pain.

**Comment by divyapola5** [comment]
Hey @jsmilani, thank you for the detailed repro steps. We are working on a fix internally.

**Comment by cipherboy** [comment]
This should be addressed in the rotation work #15277 (docs in #15238) ; feel free to test it out (`leaf_not_after_behavior` on `POST /issuer/:issuer_ref`)! This is slated for 1.11.
